modules/repository/repo.go

```diff
@@ -300,3 +313,67 @@ func PushUpdateAddTag(repo *models.Repository, gitRepo *git.Repository, tagName
 
 	return models.SaveOrUpdateTag(repo, &rel)
 }
+
+// StoreMissingLfsObjectsInRepository downloads missing LFS objects
+func StoreMissingLfsObjectsInRepository(ctx context.Context, repo *models.Repository, gitRepo *git.Repository, endpoint *url.URL) error {
+	client := lfs.NewClient(endpoint)
+	contentStore := lfs.NewContentStore()
+
+	pointerBlobs, err := lfs.SearchPointerBlobs(gitRepo)
+	if err != nil {
+		return err
+	}
+
+	for _, pointerBlob := range pointerBlobs {
+		select {
+		case <-ctx.Done():
+			log.Info("StoreMissingLfsObjectsInRepository aborted before completion")
+			return nil
+		default:
+		}
+
+		meta := &models.LFSMetaObject{Pointer: pointerBlob.Pointer, RepositoryID: repo.ID}
+		meta, err = models.NewLFSMetaObject(meta)
+		if err != nil {
+			return fmt.Errorf("StoreMissingLfsObjectsInRepository models.NewLFSMetaObject: %w", err)
+		}
+		if meta.Existing {
+			continue
+		}
+
+		log.Trace("StoreMissingLfsObjectsInRepository: LFS OID[%s] not present in repository %s", pointerBlob.Oid, repo.FullName())
+
+		exist, err := contentStore.Exists(pointerBlob.Pointer)
+		if err != nil {
+			return fmt.Errorf("StoreMissingLfsObjectsInRepository contentStore.Exists: %w", err)
+		}
+		if !exist {
+			if setting.LFS.MaxFileSize > 0 && pointerBlob.Size > setting.LFS.MaxFileSize {
+				log.Info("LFS OID[%s] download denied because of LFS_MAX_FILE_SIZE=%d < size %d", pointerBlob.Oid, setting.LFS.MaxFileSize, pointerBlob.Size)
+				continue
+			}
+
+			stream, err := client.Download(ctx, pointerBlob.Oid, pointerBlob.Size)
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					log.Info("StoreMissingLfsObjectsInRepository aborted before completion")
+					return nil
+				default:
+				}
+				log.Error("StoreMissingLfsObjectsInRepository: LFS OID[%s] failed to download: %v", pointerBlob.Oid, err)
+				continue
+			}
+			defer stream.Close()
+
+			if err := contentStore.Put(pointerBlob.Pointer, stream); err != nil {
+				if _, err2 := repo.RemoveLFSMetaObjectByOid(meta.Oid); err2 != nil {
+					return fmt.Errorf("StoreMissingLfsObjectsInRepository RemoveLFSMetaObjectByOid: %w", err2)
+				}
+				return fmt.Errorf("StoreMissingLfsObjectsInRepository contentStore.Put: %w", err)
+			}
+		}
```

Comment:
If it's exist, we may need to add a record on database?

Reply:
`models.NewLFSMetaObject` adds a record to the database if it does not exists. But I have revised the logic and not all code paths removed the model if an error occurs. The `defer` in the loop did not make sense too. I have introduced an new method to correct this and make the error handling easier.